File: app/src/androidtest/java/com/boarbeard/generator/beimax/TestMissionImpl.java
package com.boarbeard.generator.beimax;


import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.boarbeard.generator.beimax.event.Event;
import com.boarbeard.generator.beimax.event.IncomingData;

import org.junit.runner.RunWith;

import java.util.Map;

import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class TestMissionImpl {

    /**
     * Assert incoming data stays between the preferences.
     */
    public void testGenerateIncomingDataBetweenValues() {
        MissionImpl.MissionPreferences missionPreferences = new MissionImpl.MissionPreferences();
        missionPreferences.setMaxIncomingData(4);
        missionPreferences.setMinIncomingData(2);

        for (int i = 0; i < 10; i++) {
            MissionImpl missionImpl = new MissionImpl(missionPreferences);
            assertTrue(missionImpl.generateMission());

            int incomingDataCount = 0;
            for (Map.Entry<Integer, Event> entry : missionImpl.getMissionEvents().getEntrySet()) {
                if (entry.getValue() instanceof IncomingData) {
                    incomingDataCount++;
                }
            }

            assertTrue(incomingDataCount >= 2 && incomingDataCount <= 4);
        }
    }
}

File: app/src/androidtest/java/com/boarbeard/generator/beimax/TestMissionImpl.kt
package com.boarbeard.generator.beimax

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.boarbeard.generator.beimax.event.IncomingData
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class TestMissionImpl {
    /**
     * Assert incoming data stays between the preferences for incoming data.
     */
    @Test
    fun testGenerateIncomingDataBetweenValues() {
        val missionPreferences = MissionPreferences()
        missionPreferences.incomingDataRange = 2..4
        Assert.assertEquals(2, missionPreferences.getMinIncomingData())
        Assert.assertEquals(4, missionPreferences.getMaxIncomingData())

        for (i in 0..25) {
            val missionImpl = MissionImpl(missionPreferences)
            Assert.assertTrue(missionImpl.generateMission())

            val foundIncomingDataList = missionImpl.missionEvents.entrySet.filter { it.value is IncomingData }
            val incomingDataCount = foundIncomingDataList.size
            Assert.assertTrue("found $incomingDataCount instead \n $foundIncomingDataList on test $i", incomingDataCount in 2..4)
        }
    }
}